conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/ForkDetector.scala

```diff
@@ -0,0 +1,117 @@
+package tech.cryptonomic.conseil.indexer
+
+import tech.cryptonomic.conseil.indexer.ForkDetector._
+
+import cats._
+import cats.implicits._
+import com.typesafe.scalalogging.LazyLogging
+
+/** Definitions and common utilities for fork-detection */
+object ForkDetector {
+
+  /** A data type that express the of search results of a comparison by level */
+  sealed trait IdMatchResult
+
+  case object SameId extends IdMatchResult
+  case object ForkedId extends IdMatchResult
+
+  /** Provides a block identifier, given a blockchain level,
+    * i.e. how high in the chain is the block
+    *
+    * Generically provides a way to verify some chain position to verify
+    * that a forking have happened, by comparing the block between a remote
+    * and local version.
+    *
+    * - BlockId is the type identifying the block. We can't assume anything
+    *   specific here, because each blockchain may have its own type to
+    *   identify blocks uniquely
+    * - Eff is a wrapping effect we expect from the operation call
+    *   we expect that most implementation will require something like Future/IO/Try
+    *   to be returned with the value, because such operation will need to reach
+    *   across the network or on a local service/db in some way
+    * - the search is expected to be able to match the chain internal representation for his
+    *   level/height to an individual [[Long]] value, even if it's actually represented
+    *   via a different type (e.g. a custom typed wrapped like BlockHeight(Int))
+    */
+  trait SearchBlockId[Eff[_], BlockId] {
+    def searchForLevel(level: Long): Eff[BlockId]
+  }
+
+}
+
+/** Defines the algorithms to detect and identify a forking
+  * in the blockchain in comparison to the locally indexed data
+  *
+  * @param indexerSearch run the search on the local indexer
+  * @param nodeSearch run the search on the remote blockchain node
+  * @param monad a [[Monad]] provides ways to combine results in [[Eff]]
+  * @param Eq for [[BlockId]] guarantees we can compare the ids
+  */
+class ForkDetector[Eff[_]: Monad, BlockId: Eq](
+    indexerSearch: SearchBlockId[Eff, BlockId],
+    nodeSearch: SearchBlockId[Eff, BlockId]
+) extends LazyLogging {
+
+  /** Extract blocks information from local and remote data for a specific level
+    * checking if they match.
+    *
+    * @param level to check for a fork
+    * @return a match result [[IdMatchResult]]
+    */
+  def checkOnLevel(level: Long): Eff[IdMatchResult] =
+    (indexerSearch.searchForLevel(level), nodeSearch.searchForLevel(level)).mapN(
+      (indexedBlock, chainBlock) =>
+        if (indexedBlock.eqv(chainBlock)) SameId
+        else ForkedId
+    )
+
+  /** Assuming a fork was detected for a specific level (i.e. high),
```

Comment:
I don't see the calling code yet which already knows that a fork has happened. I am just assuming for now that is still to be added. TBH, I am bit confused why the initial fork detection is outside of the binary search since it is exactly the binary search which is supposed to detect whether there was a fork at all.

Reply:
I'm pushing small bites of changes as soon as they can get reviewed and are stable enough.
I want to keep under control that the code won't impact working behaviour.

The fork detection can be actually integrated too, with a small effort, in this PR.
It would only signal that a fork was detected and how far back, since the correction is not yet in place.
It would already provide a working improvement as-is. Should I proceed as such?

Reply:
As for why separating detection and search, I might be mistaken, but had the idea that the search (which is not without some cost), might be avoided if no fork is detected.

The process as I imagine it, just before downloading any new data for the cycle:
* verify if the current chain tip locally indexed is consistent with the remote node data for that level
  * if the data match, download new data
  * if the data doesn't match, run the binary search to identify the fork point, then invalidate and download new data